lib/mqtt/MqttController.js

```diff
@@ -0,0 +1,248 @@
+        }
+        if (this.password) {
+            options.password = this.password;
+        }
+        if (this.ca) {
+            options.ca = this.ca;
+        }
+        if (this.clientCert) {
+            options.cert = this.clientCert;
+        }
+        if (this.clientKey) {
+            options.key = this.clientKey;
+        }
+        return options;
+    }
+
+    /**
+     * @private
+     */
+    async connect() {
+        if (!this.client || (this.client && this.client.connected === false && this.client.reconnecting === false)) {
+            let options = this.getMqttOptions();
+            // let hassOptions = getMqttOptions(); // TODO actually implement HAss LWT
+
+            if (this.homieEnabled) {
+                options.will = {
+                    topic: this.topicPrefix + "/" + this.identifier + "/$state",
+                    payload: HomieCommonAttributes.STATE.LOST,
+                    qos: MqttCommonAttributes.QOS.AT_LEAST_ONCE,
+                    retain: true,
+                };
+            }
+
+            this.client = mqtt.connect(
+                (this.usetls ? "mqtts://" : "mqtt://") + this.server + ":" + this.port,
+                options
+            );
+            // AsyncClient is just a wrapper and it is way more convenient in some contexts
+            this.asyncClient = new asyncMqtt.AsyncClient(this.client);
+
+            this.asyncClient.on("connect", async () => {
+                Logger.info("Connected successfully to mqtt server");
+                await this.reconfigure(async () => {
+                    await this.robotHandle.configure();
+                });
+            });
+
+            this.asyncClient.on("message", async (topic, message) => {
+                if (!Object.prototype.hasOwnProperty.call(this.subscriptions, topic)) {
+                    return;
+                }
+                const msg = message.toString();
+                await this.subscriptions[topic](msg);
+            });
+
+            this.client.on("error", (e) => {
+                if (e && e.message === "Not supported") {
+                    Logger.info("Connected to non standard compliant MQTT Broker.");
+                } else {
+                    Logger.error("MQTT error:", e.toString());
+                }
+            });
+        }
+    }
+
+    /**
+     * Whether the controller is initialized. The Homie structure may only be modified while the controller is not
+     * initialized.
+     *
+     * @return {boolean}
+     */
+    isInitialized() {
+        return this.initialized;
+    }
+
+    /**
+     * @callback reconfigureCb
+     * @return {Promise<void>}
+     */
+
+    /**
+     * Set the controller in reconfiguration mode, then call the callback cb. Attributes, devices and properties can be
+     * safely reconfigured or changed  while inside the callback.
+     *
+     * @param {reconfigureCb} cb
+     * @return {Promise<void>}
+     */
+    async reconfigure(cb) {
```

Comment:
An async method that calls a callback?

Reply:
It's not fully implemented yet but it will have publish topics before and after running the callback.
The callback will also have to publish stuff. Since I need the publication to be "blocking" I will use the async API.

This method will be used as follows:

```javascript
await this.controller.reconfigure(async () => {
  this.deregisterChild(stuff);
  this.registerChild(newChild);
  newChild.configure();
});
```

In other words, similarly to python's `with` statement.

Reconfigure will mark the MQTT status as not ready for consumers, run and wait for the callback which will potentially alter the topic structure and Homie attributes, then set the status back to "ready" or something else if errors occur (I haven't decided yet what to do exactly in this case).